espressif-devcon/2023/2023.md

```diff
@@ -17,15 +17,180 @@ headingDivider: 1
 
-`async` in Rust adds two new keywords to the language, `async` & `await`, where 
+`async` in Rust adds two new keywords to the language, `async` & `await`, where
 
  - `async` defines a block or function to be asynchronous
  - `await` defines yield points _within_ an `async` block or function.
 
+## async/.await
 
-## Why use async?
+- building blocks for async blocks, closures, and functions that can be paused and yield control back to the caller
+- `async` defines blocks, closures, and functions to be asynchronous
+- `await` defines a yield point within `async` code
+- `async` code returns a value that implements the `Future` trait
 
-<!-- TODO -->
+```rust
+pub async fn say_hello(uart0: &mut Serial<UART0>) {
+  let message = "Hello World!";
+  uart0.write_bytes(message).await; // yield point here!
+  let message = "Goodbye";
+  uart0.write_bytes(message).await; // another yield point here!
+}
+```
+
+## Implementing the Future trait
+
+- asynchronous computation that can produce (even empty - ()) value
+- calling `poll()`, if the future completes, `Poll::Ready(result)` is returned
+- `Poll::Pending` is returned if future is not ready yet, `wake()` call has to be arranged to call whe the `Future` is ready to make more progress
+- the `executor` will call `poll()` again when `wake()` is called
+- the executor knows via `wake()` which Futures are ready to be `poll`ed again
+
+```rust
+pub trait Future {
+    type Output;
+
+    // Required method
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output>;
+}
+
+enum Poll<T> {
+    Ready(T),
+    Pending,
+}
+```
+
+- read data from a socket that may or may not have data available already
+- if data is available we can read them and return `Poll::Ready(data)`, if not our future is blocked
+- the `wake()` has to be registered and called when data are ready
+
+```rust
+pub struct SocketRead<'a> {
+    socket: &'a Socket,
+}
+
+impl SimpleFuture for SocketRead<'_> {
+    type Output = Vec<u8>;
+
+    fn poll(&mut self, wake: fn()) -> Poll<Self::Output> {
+        if self.socket.has_data_to_read() {
+            // The socket has data -- read it into a buffer and return it.
+            Poll::Ready(self.socket.read_buf())
+        } else {
+            // The socket does not yet have data.
+            //
+            // Arrange for `wake` to be called once data is available.
+            // When data becomes available, `wake` will be called, and the
+            // user of this `Future` will know to call `poll` again and
+            // receive data.
+            self.socket.set_readable_callback(wake);
+            Poll::Pending
+        }
+    }
+}
+```
+
+# Why use async?
+
+- Single-threaded concurrency (multitasking), doing multiple things at once without needing threads
+- Can run on a single stack, great for microcontrollers
+- Composability of async/await instead of manually writing state machines.
+
+## Busy Loop
```

Comment:
The busy loop, interrupt and async examples actually do different things? Can we make them do the same thing, i.e check if a pin is high/low and set another pin high/low.